### ISSUE TITLE
allow jumping to notes or notebooks by typing the initial letter

### ISSUE THREAD

**Issue by ericdq** [laurent22/joplin#13185]
In Joplin for desktop it is sometimes a bit complicated to jump to a specific note or notebook when we have many.

You can use the browser but it is not convenient when you are focused on the list of notebooks or notes.

I suggest allowing the use of the letters on the keyboard. Also the home and end keys should work to jump to the first and last one respectively.

In the notebook list it works with control+start or end, but in the notes list none of these combinations work.

**Comment by mrjo118** [comment]
On desktop there is a goto anything feature which you can use for this

**Comment by ericdq** [comment]
I had seen this function and although it is more convenient, for a keyboard or screen reader user it is easier if you are already in a list of notes or notebooks, to press the letter B three times if you are trying to get to the third note.

Currently you have to press control+P to display the box, type the beginning of the match, select the note with arrows and finally open it with enter.
I have also tested it with a notebook called home stuff. I had to write the first three letters to focus it. If I wrote @ and the first letter, it appeared in position 25.

I think they can be complementary and not exclusive methods.

**Comment by personalizedrefrigerator** [comment]
Supporting this is also recommended by [the WAI ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/):
>  Type-ahead is recommended for all trees, especially for trees with more than 7 root nodes:
>
>  -  Type a character: focus moves to the next node with a name that starts with the typed character.
>   -  Type multiple characters in rapid succession: focus moves to the next node with a name that starts with the string of characters typed.

**Comment by personalizedrefrigerator** [comment]
**Status update**: I've created a draft [implementation on this branch](https://github.com/personalizedrefrigerator/joplin/commit/988a4a04f92215910cdef98a23ba0c1c64ebd460). Currently:
- Pressing an alphanumeric key (or any key event that has `event.key.length === 1` ([doc: event.key](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key)) shows and focuses a text input.
- Typing in the text input moves the focus to the next matching notebook, tag, or header in the sidebar.
- If the text input's content doesn't match any notebook/tag/header, the text input's content is replaced with its last character.
- After two seconds of inactivity, the text input is hidden and focus is moved back to the note list.

The current implementation has a few problems:
- If the first character comes from a [composition event](https://developer.mozilla.org/en-US/docs/Web/API/CompositionEvent), it isn't handled. This might cause problems if, for example, searching for a notebook that starts with a CJK character.
   - One workaround is to type a letter (e.g. <kbd>a</kbd>), *then* use an input method to type the special character.
- Auto-clearing the text input after a delay may violate [the "timing adjustable" WCAG guideline (WCAG 2.2 SC 2.2.1)](https://www.w3.org/TR/WCAG22/#timing-adjustable).

A workaround for both of the above might be to show the typeahead input (and keep it open) if the user presses <kbd>ctrl</kbd>-<kbd>f</kbd> while the notebook list is selected.

**Comment by ericdq** [comment]
As a screen reader user I can't tell you if the focus becomes visible. One example I can think of right now is the one in the VisualStudioCode file browser.

I would say it works only with letters or numbers. For special symbols you could still use the go to option, or search with control+F.
I was referring to a more limited behaviour. Here you would have to see where the user has the focus. If in the list of notebooks or in the list of notes and tasks within that notebook.

Once placed on one of those items, then when the user presses the initial letter or number, jump to the first match, if he/she presses the same character again, jump to the next one starting with the same character. If he/she enters a different one, check for matches.

In this case it would not be an accumulation of characters, but if there are no matches, the record would be deleted and a new attempt would be made with the next character.

Explained in this way, perhaps it is a bit abstract, I could give examples if it is not clear or if I have not explained myself well.

Regarding the branch you created, I don't know if I could somehow test it in a real environment.

**Comment by ericdq** [comment]
With respect to criterion 2.2.1, I would apply the following exception, but I don't think it would be necessary to do something so complex for basic keyboard operation.

Essential exception: The time limit is essential and, if extended, would invalidate the activity;

**Comment by personalizedrefrigerator** [comment]
Thank you for the feedback @ericdq!

I've opened a [pull request](https://github.com/laurent22/joplin/pull/13469) with more limited functionality than the original draft implementation:
- For now, there's still no <kbd>ctrl</kbd>-<kbd>f</kbd> support.
- Only one-character typeahead is supported. Each keypress only jumps to the first matching notebook.

Additionally, for now the change only applies to the notebook list (not the note list).

**Comment by ericdq** [comment]
thanks for your work @personalizedrefrigerator I guess this will be available for testing once a new beta version is released?

I don't know if it would be very complex to provide support for the start key to go to the first element and the end key to go to the last element. Sorry, I just saw in the pr that it is already supported.

**Comment by ericdq** [comment]
in relation to the support in the list of notes and tasks, it will be left pending to see if this feature works well in the first situation, or is it because the implementation is more complex. I have seen that when there are tasks a previous text appears, which I imagine will not affect the filtering by typed characters. Or the filter has to be adjusted to omit this part of complete or incomplete task.

**Comment by personalizedrefrigerator** [comment]
Reopening as only partially resolved.

**Comment by personalizedrefrigerator** [comment]
> available for testing once a new beta version is released?

It should be available for testing when a new beta version is released.

> is it because the implementation is more complex?

It's because the implementation is more complex and modifies a different UI component. In particular, it's possible to 1) switch to a different note list style (with different information) and 2) to create plugins that show different per-item text in the note list by providing a custom rendering function. Ideally, this text would be used for the "jump to item" logic.
   - One option might be to initially only support "jump to item" for the default note list, using the `title` property of notes.

**Comment by ericdq** [comment]
I have tested the change in the new beta and it is exactly what I needed.

I have found that in notebooks that do not have subnotebooks that only have one level it does not jump to them. For example, the item all notes, or any other folder that only contains notes but no subbooks does not allow to jump to them.

**Comment by ericdq** [comment]
I think this is what you were counting on, but I'll leave it here. When the focus is on the notebook list, if we press the keys to search in all the notes, or the shortcut to go to, these commands do not work, but place the focus on the letter pressed without taking into account the rest. It occurs to me that instead of taking them into account, when detecting that we press these keys, these are omitted from the keystroke in the list of notes and act as if we were not inside that control so that the action can be performed regardless of where we have placed the focus.

**Comment by personalizedrefrigerator** [comment]
Thank you for testing!

> I have found that in notebooks that do not have subnotebooks that only have one level it does not jump to them. For example, the item all notes, or any other folder that only contains notes but no subbooks does not allow to jump to them.

For me, pressing <kbd>shift</kbd>-<kbd>A</kbd> near the end of the sidebar moves focus to "All notes". It seems that the "jump to notebook by key" doesn't work if the "Notebooks" header is selected. Could that be the issue?

> When the focus is on the notebook list, if we press the keys to search in all the notes, or the shortcut to go to, these commands do not work,

That should be fixed by #13485.

**Comment by ericdq** [comment]
If I press sift+T, in my case because I use Joplin in Spanish, it jumps to the list of all the notes. Do I have to press it because the initial is in capital letters?

If so, the filter could be modified so that regardless of whether it starts with a capital or lowercase letter, it jumps to the notebooks. This can cause problems, for example if we have capital letters enabled or we don't remember whether we start a notebook with upper or lower case.

When the< focus is on the notebook list, if we press the keys to search in all the notes, or the shortcut to go to, these commands do not work,>

That's right, that's what happened. If the first level is deployed I would have to jump to the second level.
